### PR TITLE
feat: pivot table row grouping and subtotals

### PR DESCRIPTION
**Pull request by agha4to**
Closes: https://github.com/lightdash/lightdash/issues/10267

### Description:
Adds row grouping and subtotal functionality to Pivot Tables. Pivot Table and non-Pivot Table components have distinct implementations, and ideally we'd like to combine them. This code does not do that. Instead, it keeps them separate. The underlying Pivot Table data structures have been "retrofitted" rather than rewritten at this time. I think that task is better done in the future when combining the two components, to avoid too much wasted effort.

image

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (ZeRego) [inline packages/frontend/src/components/common/PivotTable/index.tsx:165]: Yes, I think we should move this logic to `pivotQueryResults.ts` so all the data is calculated in that function and this component is only about rendering and actions.
If the outcome of a hook are values and those values are not updated anywhere else, it makes sense to be a `useMemo` and avoid multiple useState.

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (ZeRego) [inline packages/frontend/src/components/common/PivotTable/index.tsx:56]: We can export the methods from the other file and reuse them here.

**Review by ZeRego — COMMENTED** [review]
I would appreciate if you rebase this PR after the updates 🙏 🙇

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx]: - Moved data structure "retrofit" to pivotQueryResults.ts
- Changed useState to useMemo where applicable
OK, I was thinking that pivotQueryResults.ts could be reworked to remove the "retrofit" when pivot and non-pivot components are combined.

**Comment by agha4to** [comment]
@ZeRego, thank you for your patience. Refactored per your recommendations and fixed some tests. Thanks again!

**Comment by almeidabbm** [comment]
Hi @agha4to 👋 

I have been testing this and stumbled across a small bug:
I cannot change the order of the row dimensions. Don't think this is related to this PR itself, but allowing to do it would help with displaying the correct row grouping

https://github.com/user-attachments/assets/0da44252-4952-4244-8be7-ddac206df1a4

**Comment by agha4to** [comment]
Hi @almeidabbm, I've been thinking about **Rows** order and trying to implement something but I'm coming up against a few design questions.

Currently in Pivots, even without this PR:
- From the Results table, dragging columns reorders them in both the Results and Chart view
- From the pivot Configure chart sidebar, reordering items under Columns reorders the pivoted items at the top of the Chart

What's proposed:
- From the pivot Configure chart sidebar, reordering items under Rows should reorder columns in the Chart (It's somewhat confusing that reordering Rows would change column order, but it is standard for pivot tables.)

Questions:

In Configure chart, should **Rows** drags reflect in the Chart table only or also Results? A couple possibilities:

- A) Change only Chart because 1) it is after all the Chart config and 2) if dragging Rows changed Results, how would dragging Columns be handled? If we go this route, should drags of Results columns override the Configure chart Rows order and change Chart?
- B) Change both Chart and Results because it is confusing if the column orders differ between the Results and Charts. If so, again, how would dragging Columns in Chart config be handled to change Results?

What about Metrics? Shouldn't those columns also be draggable to reorder the Chart columns?

I think we need a design spec for all of this. There may be other loose ends to consider too.

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (almeidabbm) [inline packages/frontend/src/components/common/PivotTable/index.tsx:109]: @agha4to can `firstRowOnly` be typed in `pivot.ts`? I see it as `any[]` there.

This might cause issues when trying to access its properties - e.g. in this line `getField` needs a `string` argument but we might be trying to use a `column` object that doesn't have either `underlyingId` or `baseId` - seeing `packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:262`

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts
Comment (almeidabbm) [inline packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:283]: related to comment above

**Review by almeidabbm — CHANGES_REQUESTED** [review]

**Comment by almeidabbm** [comment]
> Hi @almeidabbm, I've been thinking about **Rows** order and trying to implement something but I'm coming up against a few design questions.
> 
> Currently in Pivots, even without this PR:
> 
> * From the Results table, dragging columns reorders them in both the Results and Chart view
> * From the pivot Configure chart sidebar, reordering items under Columns reorders the pivoted items at the top of the Chart
> 
> What's proposed:
> 
> * From the pivot Configure chart sidebar, reordering items under Rows should reorder columns in the Chart (It's somewhat confusing that reordering Rows would change column order, but it is standard for pivot tables.)
> 
> Questions:
> 
> In Configure chart, should **Rows** drags reflect in the Chart table only or also Results? A couple possibilities:
> 
> * A) Change only Chart because 1) it is after all the Chart config and 2) if dragging Rows changed Results, how would dragging Columns be handled? If we go this route, should drags of Results columns override the Configure chart Rows order and change Chart?
> * B) Change both Chart and Results because it is confusing if the column orders differ between the Results and Charts. If so, again, how would dragging Columns in Chart config be handled to change Results?
> 
> What about Metrics? Shouldn't those columns also be draggable to reorder the Chart columns?
> 
> I think we need a design spec for all of this. There may be other loose ends to consider too.

I see, thank you for the detailed explanation
I agree that it might take a bit more thought on how to implement this, so since this isn't exactly tied to this PR I will create a new ticket from this comment 👍

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx]: Good suggestion. I got rid of the any-s and simplified things in the process. Please let me know of any other recommendations. Thank you 👍

**Review by almeidabbm — COMMENTED** [review]
Couple things causing build errors related to types, otherwise types look good

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx:113]: Thank you. I don't know how I missed that 😞

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx]: Thanks again. Changed it up a bit to leave `item` undefined if there's no `underlyingId` or `baseId` since `item` is not needed unless those IDs are available.

**Comment by agha4to** [comment]
Found/fixed bug when drilling down when subtotals disabled and metrics as rows enabled. Header info needed to be analyzed and available to the drilldown function.

**Comment by IrakliJani** [comment]
Hi @agha4to, amazing work as always. I am almost done with the code review, but in the meantime, there’s a small feature bug that I found:

As you can see here, I have expanded the row, and the actual value is `-` but it is still aggregated as $0.00. This might confuse someone, as they might think the underlying value is 0, but in fact, it’s null.

CleanShot 2024-07-31 at 21 03 13@2x

What I'm describing is:
- if all values are null, then the sum of those values should be displayed as null.
- If all values are numbers - sum them up.
- If it’s a mix (numbers and nulls) - also sum them up.

**Comment by agha4to** [comment]
Thank you, @IrakliJani. I'll look into making the change. Note, I think we will have to also make this change in the non-pivot table code. Prior discussion: https://github.com/lightdash/lightdash/pull/9160#issuecomment-2027536401

**Comment by agha4to** [comment]
Null handling change **_not_** made in the non-pivot implementation, but please let me know if I should in this PR. Changes required there would essentially be identical to changes done here. Thank you!

**Comment by IrakliJani** [comment]
continuing review

**Comment by agha4to** [comment]
Nice, @IrakliJani. Please note the wrapping after the number in parentheses is an issue in the **non**-pivot table implementation too. So I think we have 2 things to fix there, but I don't know if we want to sneak those into this PR:
1) null handling
2) no wrap after parentheses

**Comment by IrakliJani** [comment]
@agha4to oh I see.. we can deal with that separately.

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts
Comment (IrakliJani) [inline packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:221]:
```suggestion
                (uniqueIdsForDataValueColumns[colIndex] ?? '') + 
```
should this be

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (IrakliJani) [inline packages/frontend/src/components/common/PivotTable/index.tsx:693]: iirc CellHead renders th instead of td which would be semantically inaccurate. what's the reason of using it here?

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.test.ts
Comment (agha4to) [reply packages/frontend/src/hooks/pivotTable/pivotQueryResults.test.ts:97]: Yes, using the string "undefined" which looks odd, but the idea is to have a unique key and "undefined" denotes the true value for that column. We could just use blank text `""` if undefined.

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts
Comment (agha4to) [reply packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:221]: Hah, I just suggested that above without realizing you wrote this.

File: packages/frontend/src/components/common/PivotTable/index.tsx
Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx:693]: There was a reason, but I cannot remember. Something to do with formatting. Let me see if I can figure it out.

Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx:693]: OK, I think it is because the cell in the bottom left renders in white if we use Cell:
image
With CellHead:
image

Comment (agha4to) [reply packages/frontend/src/components/common/PivotTable/index.tsx:693]: You're right about th instead of td, but the other row total "cells" also render as th:
image

**Comment by IrakliJani** [comment]
Thank you for the changes, @agha4to. I’ve finished the code review, and it looks great. Tests are all ✅ as well 🙌  !!!

There are a few minor bugs that I noticed during further testing since we switched from no-library to react-table:

- There is an unnecessary hash sign and "TH gray color" issue. I believe this is new and doesn’t look nice when there’s an empty cell next to it (see the screenshot below).
CleanShot 2024-08-13 at 14 29 29@2x

- The styling of the TH (gray-colored cells) is inconsistent. I think 1 should be white, and 2, 3 and 4 should be gray.
CleanShot 2024-08-13 at 14 20 00@2x

- The UI input for subtotals/metrics as rows is correctly displayed, but the state values are incorrect. I checked the conditions when they are enabled or not, and they are correct. However, the state should also update once certain conditions change to avoid impossible UI states
CleanShot 2024-08-13 at 14 21 32@2x

Let me know if these make sense and if you need any help with the issues! 🙌

**Comment by IrakliJani** [comment]
I found one major bug: row number appearing in the second column:
see the video so you can reproduce on your end:

https://github.com/user-attachments/assets/e6c92ff2-203d-47da-a698-9a397fb2bc0b

**Comment by agha4to** [comment]
Awesome finds, @IrakliJani! I was able to fix these two but want to test so will push code here hopefully by eow:
1) Row number in wrong column
2) Checkbox UI impossible state

And regarding these other two:

3) Hash symbol background is grayed because it's the header for row number, and consistent with non-pivot table implementation. Please lmk and I can make it white, although then we should probably change the non-pivot version.
4) Current production build has the column total zebra shading (and lack of # header btw / see below) so not sure if there is some desired logic to it. Again please lmk and I can make it all gray, except the bottom cell under #. Although, perhaps just make the whole row gray including that first cell? Whatever you say 😉 np 👍 lmk
image

**Comment by agha4to** [comment]
Here's a look at all gray column totals / bottom row:
image

Please also note that in the non-pivot version, the column totals / bottom row is all white, and the word "Total" is all the way to the left:
image

There are other variations between pivot and non-pivot implementations:
- non-pivot row number column is left-justified
- non-pivot Table name in header is not bold
- null handling, mentioned above
- no wrap after parentheses, mentioned above
- very likely more than this

**Comment by IrakliJani** [comment]
> Here's a look at all gray column totals / bottom row: image
> 
> Please also note that in the non-pivot version, the column totals / bottom row is all white, and the word "Total" is all the way to the left: image
> 
> There are other variations between pivot and non-pivot implementations:
> 
>     * non-pivot row number column is left-justified
> 
>     * non-pivot Table name in header is not bold
> 
>     * null handling, mentioned above
> 
>     * no wrap after parentheses, mentioned above
> 
>     * very likely more than this

thanks @agha4to 

It's fine as it is now. no need to have parity between the two at the moment.

I created a ticket based on your comment here: https://github.com/lightdash/lightdash/issues/11187 feel free to add more there 🙌

**Comment by IrakliJani** [comment]
> Awesome finds, @IrakliJani! I was able to fix these two but want to test so will push code here hopefully by eow:
> 
>     1. Row number in wrong column
> 
>     2. Checkbox UI impossible state
> 
> 
> And regarding these other two:
> 
>     3. Hash symbol background is grayed because it's the header for row number, and consistent with non-pivot table implementation. Please lmk and I can make it white, although then we should probably change the non-pivot version.
> 
>     4. Current production build has the column total zebra shading (and lack of # header btw / see below) so not sure if there is some desired logic to it. Again please lmk and I can make it all gray, except the bottom cell under #. Although, perhaps just make the whole row gray including that first cell? Whatever you say 😉 np 👍 lmk
>        image

1 & 2 awesome! 

3. A hash is not a requirement, and we don't have to be consistent with the non-pivot table. I don't think it plays any role or adds any visual aid, especially when pivoting on multiple levels.

4. You can ignore it if it's the same as the prod build. The idea here is to have no styling (white background + blank cell) on empty cells that have no value.

For example:
- A footer row with a total should be gray.
- A footer row with a null total should still be gray.
- A footer row with a label should still be gray.
- A footer row that displays nothing should be white.

The same goes for column totals, the index column, and headers.

@agha4to Did you push the code? I can't see a new commit to check the diff.

**Comment by agha4to** [comment]
QA should be ✅ tomorrow I think. Will update w/ screenshots and we can go from there!
Thanks for https://github.com/lightdash/lightdash/issues/11187; would be nice to combine Pivot and non-Pivot implementations to avoid discrepancies
Thank you, @IrakliJani.

**Comment by IrakliJani** [comment]
thanks @agha4to 🙌

**Comment by agha4to** [comment]
Came across another possible issue with checkbox state so looking into that.

Regarding UI though:

1) I left the `#` but we can certainly change it. It may help users answer with "What are these numbers in the first column?" but then `#` is not an international symbol for "number" so perhaps confusing to some.
image

---

2) I think the white cell shown below between `#` and `Payments Total revenue` helps distinguish between headers referring to columns and headers referring to pivoted dimensions. Perhaps headers for pivoted items should have a colon after them like `Orders Order date year:` or other visual cue besides just right alignment.
image

---

3) Bottom row shading may still need some changes:
image

**Comment by agha4to** [comment]
It turns out the 2 issues from QA are unrelated to any changes from this PR:
1) Pivot table does not update after filter(s) applied and query run
2) "Show row totals" & "Show column totals" checkboxes change values unexpectedly, due to mixup [here](https://github.com/lightdash/lightdash/blob/main/packages/frontend/src/components/VisualizationConfigs/TableConfigPanel/GeneralSettings.tsx#L71-L77) which looks like an easy fix btw

**Comment by IrakliJani** [comment]
@agha4to sorry for the long pause.. we didn’t have the capacity to give a final review and a QA

Is there anything remaining before we approve the code changes?

here are minor UI issues I noticed during QA:

CleanShot 2024-09-25 at 12 53 55@2x

**Comment by IrakliJani** [comment]
render link: https://lightdash-pr-11681.onrender.com

**Comment by agha4to** [comment]
I heard y'all were busy. I don't think there's a rush--no problem. Afaik, no remaining issues. Here's a look at the recent shading changes:

1. Last row of Row # column is **no longer** shaded
2. Last row of dimension columns are **no longer** shaded
3. Last row of metric columns that do not have column totals are **no longer** shaded
4. Row total columns, not including last row, **are** shaded

image

**Review by IrakliJani — APPROVED** [review]

**Comment by IrakliJani** [comment]
@agha4to approving code changes ✅ great work 🙌

**Review by almeidabbm — APPROVED** [review]

**Pull request merged**
